### PR TITLE
Use a map instead of list, to keep track of enforcer resources within adapter

### LINKED ISSUE #2135: Application Key Mappings are never removed.
### Description:
Application Key Mappings are never removed from the adapter's xds cache as adapter does not respond for the Remove_Application_Key event. 

### Steps to reproduce:
Delete application from api manager and observe the adapter logs.

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### LINKED ISSUE #2136: Improve the subscription, application data store mechanism in adapter
### Describe your problem(s)
At the moment, lists are used to keep the application, application key mappings, subscripitons and throttle policies. It would be far more efficient if the maps are used instead of lists.

### How will you implement it

---

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
$subject. This would improve the adding, and removing enforcer resources when JMS events are received. Additionally, the duplicate `create` resource events would not impact the expected behavior of adapter.

In addition, this would fix the issue where application key mappings are not removed when REMOVED_KEY_MAPPING event is recieved.

### Issues

Fixes #2135
Fixes #2136 

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by renuka-fernando — APPROVED** [review]

File: adapter/internal/messaging/notification_listener.go [hunk 1]
```diff
@@ -282,16 +287,16 @@ func handleApplicationEvents(data []byte, eventType string) {
...
-			eh.AppList.List = append(eh.AppList.List, application)
-			logger.LoggerInternalMsg.Infof("Application %s is added.", applicationEvent.ApplicationName)
+			eh.ApplicationMap[application.UUID] = &application
+			logger.LoggerInternalMsg.Infof("Application %s is added.", applicationEvent.UUID)
 		} else if applicationEvent.Event.Type == applicationUpdate {
-			eh.AppList.List = removeApplication(eh.AppList.List, applicationEvent.ApplicationID)
-			eh.AppList.List = append(eh.AppList.List, application)
-			logger.LoggerInternalMsg.Infof("Application %s is added.", applicationEvent.ApplicationName)
+			eh.ApplicationMap[application.UUID] = &application
+			logger.LoggerInternalMsg.Infof("Application %s is updated.", applicationEvent.ApplicationName)
```
Comment (renuka-fernando) [inline adapter/internal/messaging/notification_listener.go:294]: For, create and delete applications we are logging UUID, is there any reason to log ApplicationName for update events?

File: adapter/internal/messaging/notification_listener.go [hunk 2]
```diff
@@ -318,14 +323,13 @@ func handleSubscriptionEvents(data []byte, eventType string) {
...
-		updateSubscription(subscriptionEvent.SubscriptionID, sub)
+		eh.SubscriptionMap[sub.SubscriptionID] = &sub
 	} else if subscriptionEvent.Event.Type == subscriptionUpdate {
-		eh.SubList.List = removeSubscription(eh.SubList.List, subscriptionEvent.SubscriptionID)
-		updateSubscription(subscriptionEvent.SubscriptionID, sub)
+		eh.SubscriptionMap[sub.SubscriptionID] = &sub
 	} else if subscriptionEvent.Event.Type == subscriptionDelete {
-		eh.SubList.List = removeSubscription(eh.SubList.List, subscriptionEvent.SubscriptionID)
+		delete(eh.SubscriptionMap, sub.SubscriptionID)
 	}
```
Comment (renuka-fernando) [inline adapter/internal/messaging/notification_listener.go:331]: How about adding logs for other events as well? same, that have already done for the application events.

Comment (VirajSalaka) [reply adapter/internal/messaging/notification_listener.go]: fixed with https://github.com/wso2/product-microgateway/pull/2126/commits/618511c91e7788869966c069893c6ec4acb27919

File: adapter/internal/messaging/notification_listener.go (hunk 1 above)
Comment (VirajSalaka) [reply adapter/internal/messaging/notification_listener.go]: fixed with https://github.com/wso2/product-microgateway/pull/2126/commits/618511c91e7788869966c069893c6ec4acb27919

File: adapter/internal/messaging/notification_listener.go [hunk 3]
```diff
@@ -362,24 +373,21 @@ func handlePolicyEvents(data []byte, eventType string) {
 		logger.LoggerInternalMsg.Infof("Policy: %s for policy type: %s", policyEvent.PolicyName, policyEvent.PolicyType)
 	}
 
-	// TODO: (VirajSalaka) Decide if it is required to have API Level Policies
-	// if strings.EqualFold(apiEventType, policyEvent.PolicyType) {
-	// 	var apiPolicyEvent APIPolicyEvent
-	// 	json.Unmarshal([]byte(string(data)), &apiPolicyEvent)
-	// } else
```
Comment (Rajith90) [inline adapter/internal/messaging/notification_listener.go:369]: Shall we remove the unwanted code

**Review by Rajith90 — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] dev-deployment : 20210817.1

**Comment by choreo-cicd** [comment]
[failed] stage-deployment : 20210817.4

**Comment by choreo-cicd** [comment]
[succeeded] stage-deployment : 20210817.5

**Comment by choreo-cicd** [comment]
[succeeded] prod-deployment : 20210818.2

**Pull request merged**